daemon/logger/jsonfilelog/read.go

```diff
@@ -174,6 +181,17 @@ func followLogs(f *os.File, logWatcher *logger.LogWatcher, notifyRotate chan int
 		fileWatcher.Close()
 	}()
 
+	logWatcherCtx, logWatcherClosed := context.WithCancel(ctx)
+	go func() {
+		select {
+		case <-logWatcher.WatchClose():
+			fileWatcher.Remove(name)
```

Comment:
Sorry, it is a little late over here. I've just been having a think about the order of execution. If a `fileWatcher.Remove` happens after `fileWatcher.Close` it would likely hang as `Remove` would be waiting for an `IN_IGNORE` event but `readEvents` (in https://github.com/docker/docker/blob/362eb4cfbb09b5dc0fe648360a2e6d0546bf6613/vendor/src/gopkg.in/fsnotify.v1/inotify.go) would have exited because the watcher was closed so it would never receive it. This makes me a little concerned about the `Remove` after `ctx.Done()` as the `fileWatcher.Close()` in the second defer could have run by the time `Remove` is called.

I'm fairly new to Go so if that doesn't sounds right (could also be because I'm tried) let me know, otherwise I might switch to a `fileWatcher.Remove` before the `Close` in the defer above the go block. If it has already been removed it should have no issue, it will just return an error that we ignore anyway: https://github.com/docker/docker/blob/362eb4cfbb09b5dc0fe648360a2e6d0546bf6613/vendor/src/gopkg.in/fsnotify.v1/inotify.go#L135